### PR TITLE
Starter class with abundance features

### PR DESCRIPTION
**Pull request by camisowers**
**If you haven't already, please read through our [contributing guidelines](https://ark-analysis.readthedocs.io/en/latest/_rtd/contributing.html) before opening your PR**

**What is the purpose of this PR?**

This PR includes starter code for the SpaceCat functionality. It addresses part of #1, but doesn't include any features other than the cell abundances (densities and density ratios).

Will merge onto a working branch `spacecat_v1` instead of main.

Currently, there's some example code in example_SpaceCat.ipynb on how to convert a csv cell table to anndata, example of running the SpaceCat feature generation, and saving files to csvs so the user can know where the feature dfs are stored in the anndata object. This notebook can be expanded with additional written instructions for the user, or ported into the readme if we decide that's a better place for it.

**How did you implement your changes**

Laid out in #1. 
* preprocess.py contains functions for preparing the cell table for SpaceCat processing (i.e. adding functional marker positivity to the table, later will include distance matrix generation and more)
* features.py contains the FeatureSpace() class, which does all the heavy lifting for generating the intermediate stats tables and the final feature table
* example_SpaceCat.ipynb has example code for the user to understand how the class works and where the output is stored

**Remaining issues**

- [x] some more generalization of parameters
- [x] add notebook and/or example code to the readme
- add some tests in a later PR
- port notebook code to readme

### CODE REVIEW

File: src/SpaceCat/features.py [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+            pd.DataFrame: long format dataframe containing the summarized data
+        """
+
+        verify_in_list(cell_type_col=cluster_col_name, cell_table_columns=table.columns)
+        verify_in_list(drop_cols=drop_cols, cell_table_columns=table.columns)
+        if subset_col:
+            verify_in_list(subset_col=subset_col, cell_table_columns=table.columns)
+
+        # drop columns from table
+        table_small = table.loc[:, ~table.columns.isin(drop_cols)]
```
Comment (alex-l-kong) [inline src/SpaceCat/features.py:67]: Alternatively: `table_small = table.drop(columns=drop_cols, errors="ignore")` will bypass errors if a name in `drop_cols` doesn't exist.

File: src/SpaceCat/features.py [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+            if normalize:
+                transformed = grouped_table.agg('mean')
+            else:
+                transformed = grouped_table.agg('sum')
+            transformed.reset_index(inplace=True)
+
+            # reshape to long df
+            long_df = pd.melt(
+                transformed, id_vars=[self.image_col, cluster_col_name], var_name=var_name)
+            long_df = long_df.rename(columns={cluster_col_name: 'cell_type'})
```
Comment (alex-l-kong) [inline src/SpaceCat/features.py:97]: The logic between the `if` and `else` clauses are different enough that I'd argue for making two separate helper functions.

File: src/SpaceCat/features.py [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+        Returns:
+            pd.DataFrame:
+                long format dataframe containing the summarized data
+        """
+
+        # first generate df without subsetting
+        drop_cols_all = drop_cols.copy()
+        if subset_col is not None:
+            drop_cols_all = drop_cols_all + [subset_col]
```
Comment (alex-l-kong) [inline src/SpaceCat/features.py:128]:
```suggestion
        drop_cols_all = drop_cols + [subset_col] if subset_col is not None else drop_cols.copy()
```

File: src/SpaceCat/features.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+                transformed, id_vars=[self.image_col, cluster_col_name], var_name=var_name)
+            long_df = long_df.rename(columns={cluster_col_name: 'cell_type'})
+
+        if subset_col:
+            long_df = long_df.rename(columns={subset_col: 'subset', self.label_col: 'value'})
+
+        return long_df
+
+    def create_long_df(self, table, cluster_col_name, result_name, var_name, subset_col=None,
+                       cluster_stats=False, normalize=False, drop_cols=[],
```
Comment (alex-l-kong) [inline src/SpaceCat/features.py:105]: Generally bad practice to pass in mutable types as default arguments. Setting `drop_cols=None` would be better.

File: src/SpaceCat/features.py [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+        grouped_cell_freq_region = cell_table_clusters[[self.image_col, self.compartment_col]].\
+            groupby([self.image_col], observed=True)
+        grouped_cell_freq_region = grouped_cell_freq_region[self.compartment_col].\
+            value_counts(normalize=True)
+        grouped_cell_freq_region = pd.DataFrame(grouped_cell_freq_region)
+        grouped_cell_freq_region.columns = ['value']
+        grouped_cell_freq_region.reset_index(inplace=True)
+        grouped_cell_freq_region['metric'] = 'total_cell_freq'
+        grouped_cell_freq_region.rename(columns={self.compartment_col: 'subset'}, inplace=True)
+        grouped_cell_freq_region['cell_type'] = 'all'
```
Comment (alex-l-kong) [inline src/SpaceCat/features.py:208]: I'd argue for breaking down these three individual steps into helper functions.

File: src/SpaceCat/features.py [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+
+        count_df = total_df_clusters.loc[total_df_clusters.metric.isin(count_metrics), :]
+        compartment_area_df = compartment_area_df.rename(columns={self.compartment_col: 'subset'})
+        all_area = compartment_area_df[[self.image_col, self.compartment_col + '_area']].groupby(
+            by=[self.image_col]).sum().reset_index()
+        all_area['subset'] = 'all'
+        area_df = pd.concat([compartment_area_df, all_area])
+        count_df = count_df.merge(area_df, on=[self.image_col, 'subset'], how='left')
+        count_df['value'] = count_df['value'] / count_df[self.compartment_col + '_area']
+        count_df['value'] = count_df['value'] * 1000
```
Comment (alex-l-kong) [inline src/SpaceCat/features.py:230]: This step can also be broken down into a helper function. The `generate_cluster_stats` function is large right now, using helper functions will reduce code and help with flow.

File: src/SpaceCat/features.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+        Returns:
+            pd.DataFrame:
+                table with feature metadata columns included
+        """
+        # add useful metadata about the feature
+        if compartment == 'all':
+            compartment_df['feature_name_unique'] = compartment_df['feature_name']
+        else:
+            compartment_df['feature_name_unique'] = compartment_df['feature_name'] + '__' + \
+                                                    compartment
```
Comment (alex-l-kong) [inline src/SpaceCat/features.py:259]: If you really wanted to:

```suggestion
        compartment_df["feature_name_unique"] = compartment_df["feature_name"]
        if compartment != "all":
            compartment_df["feature_name_unique"] = compartment_df["feature_name_unique"] + "__" + compartment
```

**Review by alex-l-kong — CHANGES_REQUESTED** [review]
General coding practice comments to start off with.

File: src/SpaceCat/features.py (hunk 5 above)
Comment (camisowers) [reply src/SpaceCat/features.py]: Since this is the only instance the code is used, I just extracted the repetitive metadata assignment into a nested helper function.

File: src/SpaceCat/features.py [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+from alpineer.misc_utils import verify_in_list
+
+
+class SpaceCat:
+
+    def __init__(self, adata_table, image_col, label_col, cluster_columns, compartment_col):
+        self.adata_table = adata_table.copy()
+        self.image_col = image_col
+        self.label_col = label_col
+        self.cluster_columns = cluster_columns
```
[the remark this answers is not in the document]
Comment (jranek) [reply src/SpaceCat/features.py:15]: It's also unclear that area for each compartment must have the same name as the compartment_col. This would be simpler if we provide the functions to compute compartments and ensure the naming convention is correct. Else, I think it needs to be more explicitly stated

File: src/SpaceCat/features.py [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+            cluster_dfs.append(self.create_long_df(table=cell_table_clusters,
+                                                   cluster_col_name=cluster_col_name,
+                                                   result_name=result_name,
+                                                   var_name='cell_type',
+                                                   subset_col=self.compartment_col,
+                                                   cluster_stats=True,
+                                                   normalize=normalize,
+                                                   drop_cols=drop_cols))
+
+        def add_metadata(panda_series, metric, compartment_col=False):
```
Comment (jranek) [inline src/SpaceCat/features.py:220]: I might rename this to `add_feature_metadata` in case people think it's actual metadata

File: src/SpaceCat/example_SpaceCat.ipynb [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+   "id": "0d3e3f0b-6d3b-4fb9-b68b-ac004e38a86a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Save finalized tables to csv \n",
+    "os.makedirs(os.path.join(output_dir, 'SpaceCat'), exist_ok=True)\n",
+    "\n",
+    "adata_processed.uns['combined_feature_data'].to_csv(os.path.join(output_dir, 'SpaceCat', 'combined_feature_data.csv'))\n",
+    "adata_processed.uns['combined_feature_data_filtered'].to_csv(os.path.join(output_dir, 'SpaceCat', 'combined_feature_data_filtered.csv'))\n",
+    "adata_processed.uns['feature_metadata'].to_csv(os.path.join(output_dir, 'SpaceCat', 'feature_metadata.csv'))\n",
```
Comment (jranek) [inline src/SpaceCat/example_SpaceCat.ipynb:140]: Is there a reason to have this saved separately? It's already included in the combined_feature_data, right?

File: src/SpaceCat/features.py [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+                the anndata table with all intermediate and final tables appended
+        """
+        # Generate counts and proportions of cell clusters per FOV
+        cluster_params = []
+        for column in self.cluster_columns:
+            cluster_params.append([column + '_freq', column, True])
+            cluster_params.append([column + '_count', column, False])
+
+        # subset table for cluster data
+        cell_table_clusters = self.adata_table.obs[
```
Comment (jranek) [inline src/SpaceCat/features.py:466]: The naming convention is a bit hard to follow, I think it could benefit from more descriptions in text or within logging.info statements. Computing frequency and total abundance of cell types, computing the density, computing the ratios, etc. Abundance is used a bunch of times, but it's hard to follow since they're different types of abundances

File: src/SpaceCat/features.py [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+        exclude_df = pd.DataFrame({'feature_name_unique': exclude_list})
+        self.excluded_features = exclude_df
+        self.adata_table.uns['excluded_features'] = exclude_df
+        feature_df_filtered = \
+            feature_df.loc[~feature_df.feature_name_unique.isin(
+                exclude_df.feature_name_unique.values), :]
+
+        return feature_df_filtered.reset_index(drop=True)
+
+    def combine_features(self, correlation_filtering=0.7):
```
Comment (jranek) [inline src/SpaceCat/features.py:406]: This `correlation_filtering` param should be an init value for the user to change if they so choose. If False, then doesn't compute, else whatever value is set ensuring correct range

File: src/SpaceCat/example_SpaceCat.ipynb [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    "adata_processed.uns['excluded_features'].to_csv(os.path.join(output_dir, 'SpaceCat', 'excluded_features.csv'))"
```
Comment (jranek) [inline src/SpaceCat/example_SpaceCat.ipynb:141]: Would be helpful if excluded_features had a column that indicated where/why it was excluded. For example, correlation, density etc

**Review by jranek — CHANGES_REQUESTED** [review]
Looks awesome, super quick to run too :). Here are some initial thoughts for clarity

File: src/SpaceCat/example_SpaceCat.ipynb [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    "           'CD45RB', 'TIM3','IDO', 'LAG3', 'CD38', 'Fe']\n",
+    "\n",
+    "cell_data_cols = ['fov', 'label', 'cell_meta_cluster', 'cell_cluster', 'cell_cluster_broad', 'compartment', 'compartment_area', 'area', 'cell_size', 'centroid_dif', 'convex_area', 'convex_hull_resid', \n",
+    "                 'eccentricity', 'equivalent_diameter', 'major_axis_equiv_diam_ratio', 'major_axis_length',\n",
+    "                 'major_minor_axis_ratio', 'minor_axis_length', 'nc_ratio', 'num_concavities', 'perim_square_over_area',\n",
+    "                 'perimeter']\n",
+    "centroid_cols = ['centroid-0', 'centroid-1']\n",
+    "\n",
+    "adata = anndata.AnnData(cell_table.loc[:, markers])\n",
+    "adata.obs = pd.merge(cell_table.loc[:, cell_data_cols, ], on = ['fov'])\n",
```
[the remark this answers is not in the document]
Comment (camisowers) [reply src/SpaceCat/example_SpaceCat.ipynb:65]: Yeah this is where I meant to merge the Patient_ID and other metadata, nice catch!

File: src/SpaceCat/example_SpaceCat.ipynb (hunk 10 above)
Comment (camisowers) [reply src/SpaceCat/example_SpaceCat.ipynb:140]: It is saved in combined_feature_data yes, but this is a much simpler output for users to look at and understand since it explains only unique features (e.g ~500 rows) rather than having the each feature repeated as a row for each image (~500,000 rows).

File: src/SpaceCat/features.py (hunk 12 above)
Comment (camisowers) [reply src/SpaceCat/features.py:406]: Agreed I remembered to include this when doing the functional marker stats but hadn't ported it over yet!

File: src/SpaceCat/features.py (hunk 11 above)
Comment (camisowers) [reply src/SpaceCat/features.py:466]: I'll add some clarifying comments!

File: src/SpaceCat/example_SpaceCat.ipynb (hunk 10 above)
Comment (jranek) [reply src/SpaceCat/example_SpaceCat.ipynb:140]: ah gotcha!

File: src/SpaceCat/example_SpaceCat.ipynb [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+   "source": [
+    "# Preprocessing (only needs to be run once)\n",
+    "functional_marker_thresholds = [['Ki67', 0.002], ['CD38', 0.004], ['CD45RB', 0.001], ['CD45RO', 0.002],\n",
+    "                                ['CD57', 0.002], ['CD69', 0.002], ['GLUT1', 0.002], ['IDO', 0.001],\n",
+    "                                ['LAG3', 0.002], ['PD1', 0.0005], ['PDL1', 0.001],\n",
+    "                                ['HLA1', 0.001], ['HLADR', 0.001], ['TBET', 0.0015], ['TCF1', 0.001],\n",
+    "                                ['TIM3', 0.001], ['Vim', 0.002], ['Fe', 0.1]]\n",
+    "\n",
+    "adata = anndata.read_h5ad(os.path.join(output_dir, 'adata', 'adata_table.h5ad'))\n",
+    "adata_processed = preprocess_table(adata, functional_marker_thresholds)\n",
```
[the remark this answers is not in the document]
Comment (camisowers) [reply src/SpaceCat/example_SpaceCat.ipynb:86]: I think we can just convert any sparse adata.X input to dense matrices for now, and if we have issues with storage we can revisit how the preprocessing functions work and maybe adapt them to work with sparse input.

File: src/SpaceCat/features.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (camisowers) [reply src/SpaceCat/features.py:15]: For now I'll add in some if statements to ensure the image wide features get computed without issue and skip compartment specific stuff. Then we can always add in some preprocessing functions to take in masks and assign cells to individual compartments / compute the areas later!

File: src/SpaceCat/example_SpaceCat.ipynb (hunk 15 above)
[the remark this answers is not in the document]
Comment (camisowers) [reply src/SpaceCat/example_SpaceCat.ipynb:86]: > Creating the single positive and double positive functional dfs throws a lot of frame insert fragmented df warnings. I think this can be circumvented if you append to a new df/ dictionary rather than doing the thresholding directly inside the existing dataframe

I don't get there warnings. 🤔 For now I'll leave the functions as they are, but once we have package versions set I'll try it again and see if it pops up!

File: src/SpaceCat/example_SpaceCat.ipynb (hunk 13 above)
Comment (camisowers) [reply src/SpaceCat/example_SpaceCat.ipynb:146]: Currently this is only correlation based filtered features, but as we incorporate more filtering for functional markers I'll think about keeping a running list with descriptions.

**Review by jranek — APPROVED** [review]
Looks good!!

**Review by ngreenwald — APPROVED** [review]
This was a ton of work! Great start
